tap_nice_incontact/streams.py

```diff
@@ -604,6 +609,200 @@ def get_records(self,
+            else:
+                # Job status can also be: Failed, Cancelled, and Expired.
+                # https://help.nice-incontact.com/content/recording/dataextractionapi.htm
+                raise ValueError("Data extraction job failure", job_details)
+
+        records = self.download_csv(job_details['result']['url'])
+        return records
+
+    def sync(self,
+             state: dict,
+             stream_schema: dict,
+             stream_metadata: dict,
+             config: dict,
+             transformer: Transformer) -> dict:
+        """
+        The sync logic for an incremental stream.
+
+        :param state: A dictionary representing singer state
+        :param stream_schema: A dictionary containing the stream schema
+        :param stream_metadata: A dictionnary containing stream metadata
+        :param config: A dictionary containing tap config data
+        :param transformer: A singer Transformer object
+        :return: State data in the form of a dictionary
+        """
+        
+
+        start_date = singer.get_bookmark(state,
+                                        self.tap_stream_id,
+                                        self.replication_key,
+                                        config['start_date'])
+        bookmark_datetime = singer.utils.strptime_to_utc(start_date)
+        max_datetime = bookmark_datetime        
+        
+        with metrics.record_counter(self.tap_stream_id) as counter:
+            for record in self.get_records(config, bookmark_datetime):
+                # Convert data keys for CSV outputs to the correct field format
+                record = convert_data_keys(record)
+
+                if self.convert_data_types:
+                    record = convert_data_types(record, stream_schema)
+
+                transformed_record = transformer.transform(record, stream_schema, stream_metadata)
+                # pylint: disable=line-too-long
+                record_datetime = singer.utils.strptime_to_utc(transformed_record[self.replication_key])
+                if record_datetime >= bookmark_datetime:
+                    singer.write_record(self.tap_stream_id, transformed_record)
+                    counter.increment()
+                    max_datetime = max(record_datetime, bookmark_datetime)
+
+            bookmark_date = singer.utils.strftime(max_datetime)
+
+        state = singer.write_bookmark(state,
+                                    self.tap_stream_id,
+                                    self.replication_key,
+                                    bookmark_date)
+        singer.write_state(state)
+        return state
+
+    def get_records(self,
+                    config: dict = None,
+                    bookmark_datetime: datetime = None,
+                    is_parent: bool = False) -> Iterator:
+        if config.get('periods'):
+            period = config.get('periods', {}).get(self.tap_stream_id)
+        else:
+            period = self.default_period
```

Comment:
When periods is defined but the tap stream_id DNE, it will return None and the generated dates will return an empty list.

```suggestion
        periods: dict = config.get('periods') or {} 
        period = self.default_period
        if periods and periods.get(self.tap_stream_id):
            period = periods[self.tap_stream_id]
```

Reply:
Hmm good point, I'll need to change every other instance we load the periods config (there are multiple streams that use this period logic).

I'm guessing the previous behavior just assumed the config would always have these streams